proxy.go

```diff
@@ -207,23 +207,22 @@ func executeWithRetry(
 	startTime := time.Now()
 	var since float64
 
-	// keep the request body
-	body, err := io.ReadAll(req.Body)
-	req.Body.Close()
+	// Use readAndRestoreRequestBody to read the entire request body into a byte slice,
+	// and to restore req.Body so that it can be reused later in the code.
+	body, err := readAndRestoreRequestBody(req)
 	if err != nil {
-		since = time.Since(startTime).Seconds()
-
+		since := time.Since(startTime).Seconds()
 		return since, err
 	}
 
 	numRetry := 0
 	for {
-		// update body
+		rp(rw, req)
+
+		// Restore req.Body after it's consumed by 'rp' for potential reuse.
 		req.Body = io.NopCloser(bytes.NewBuffer(body))
 		req.Body.Close()
```

Comment:
why do you have a call to Close() here and there? https://github.com/ContentSquare/chproxy/pull/397/files#diff-8b7f25bac2ae04c2ce2abb628c6aa28103fd113103a62f3e9d232fa995684915R111

I don't know what should be done, but it doesn't look logic to have inconsistencies for the same use case.

Reply:
same comment for https://github.com/ContentSquare/chproxy/pull/397/files#diff-8b7f25bac2ae04c2ce2abb628c6aa28103fd113103a62f3e9d232fa995684915R317

Reply:
You are right. This one was at the beginning. I'll check if can be removed

Reply:
req.Body.Close() appears in few places. Maybe, better to refactor it in a separate PR?